lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,256 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/TT/IR/TT.h"
+#include "ttmlir/Dialect/TTIR/Transforms/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/SmallVector.h"
+
+#include "llvm/ADT/SmallSet.h"
+
+namespace mlir::tt::ttir {
+#define GEN_PASS_DEF_TTIRHOISTTRANSFORM
+#include "ttmlir/Dialect/TTIR/Transforms/Passes.h.inc"
+
+//===----------------------------------------------------------------------===//
+// Hoist CPU ops to standalone funcs pass
+//===----------------------------------------------------------------------===//
+
+// helper function to get ranks of an op's operands
+// we use this to populate attrs which we need to tensor unpacking operations
+// later
+static llvm::SmallVector<int64_t, 4> getTensorRanks(mlir::Operation *op) {
+  llvm::SmallVector<int64_t, 4> ranks;
+
+  // Iterate over operands (inputs)
+  for (auto operand : op->getOperands()) {
+    // Check if the operand is a tensor
+    if (auto tensorType = dyn_cast<mlir::RankedTensorType>(operand.getType())) {
+      // Add the rank of the tensor (number of dimensions)
+      ranks.push_back(tensorType.getRank());
+    }
+  }
+
+  return ranks;
+}
+
+// generate unique name base on operation type + argument tensors dims & types
+static llvm::SmallString<16> generateHoistedFuncName(mlir::Operation *op) {
```

Comment:
This might end up generating 2 equal names if the parameters are equal, should we append a rand string to fn name to "guarantee" uniqueness?

Reply:
That's right, but I think that's desirable.  Below, I do a check to see if a func of the name already exists, and if so skip emitting new func, and simply call the same func, to avoid generating duplicate functions.  In my testcase, I exercise this logic as well (2 of the 4 calls are same signature, only 3 funcs are generated)